pkg/service/worker_service.go

```diff
@@ -0,0 +1,187 @@
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/keepalive"
+	"google.golang.org/grpc/status"
+)
+
+// WorkerService is the gRPC Service of Worker
+type WorkerService struct {
+	proto.UnimplementedWorkerServiceServer
+	jobsManager job.JobsManager
+}
+
+// newWorkerService initiates new service
+func newWorkerService() (*WorkerService, error) {
+	jobsManager, err := job.NewManager()
+	if err != nil {
+		return nil, err
+	}
+
+	return &WorkerService{
+		jobsManager: jobsManager,
+	}, nil
+}
+
+// StartJob starts a job corresponding to user request
+func (service *WorkerService) StartJob(ctx context.Context, cmd *proto.Command) (*proto.Job, error) {
+	if service.jobsManager == nil {
+		return nil, status.Errorf(codes.Unavailable, "Job Managers is not ready")
+	}
+
+	// TODO Get Owner common name from certificate
+	jobID, err := service.jobsManager.CreateJob(cmd.Cmd, cmd.Args, "User CN")
+	if err != nil {
+		return nil, err
+	}
+
+	return &proto.Job{
+		Id: jobID,
+	}, nil
+}
+
+// StopJob terminates a job specified by user request
+func (service *WorkerService) StopJob(ctx context.Context, request *proto.StopRequest) (*proto.JobStatus, error) {
+	if service.jobsManager == nil {
+		return nil, status.Errorf(codes.Unavailable, "Job Managers is not ready")
+	}
+
+	j, ok := service.jobsManager.GetJob(request.Job.Id)
+	if !ok {
+		badRequest := &errdetails.BadRequest{
+			FieldViolations: []*errdetails.BadRequest_FieldViolation{
+				{
+					Field:       "job",
+					Description: "Job ID is not correct",
+				},
+			},
+		}
+
+		return nil, job.ReportError(codes.NotFound, "Fail to stop job", badRequest)
+	}
+
+	if err := j.Stop(request.Force); err != nil {
+		return nil, err
+	}
+
+	return j.Status(), nil
+}
+
+// QueryJob returns the status of a job
+func (service *WorkerService) QueryJob(ctx context.Context, protoJob *proto.Job) (*proto.JobStatus, error) {
+	if service.jobsManager == nil {
+		return nil, status.Errorf(codes.Unavailable, "Job Managers is not ready")
+	}
+
+	// TODO Get Owner common name from certificate
+	j, ok := service.jobsManager.GetJob(protoJob.Id)
+	if !ok {
+		badRequest := &errdetails.BadRequest{
+			FieldViolations: []*errdetails.BadRequest_FieldViolation{
+				{
+					Field:       "job",
+					Description: "Job ID is not correct",
+				},
+			},
+		}
+
+		return nil, job.ReportError(codes.NotFound, "Fail to query job", badRequest)
+	}
+
+	return j.Status(), nil
+}
+
+// StreamLog maintains a stream of job logs specified by user
+/*
+func (service *WorkerService) StreamLog(job *proto.Job, stream proto.WorkerService_StreamLogServer) error {
+	return fmt.Errorf("Unimplemented")
+}
+*/
+
+// Cleanup cleanups the service
+func (service *WorkerService) Cleanup() error {
+	if service.jobsManager == nil {
+		return fmt.Errorf("Job Managers is not initiated")
+	}
+
+	return service.jobsManager.Cleanup()
+}
+
+// WorkerServer is gRPC Server of worker service
+type WorkerServer struct {
```

Comment:
can this be combined with `WorkerService`?

Reply:
Yes, it can, but I don't think it should be. Because the WorkerServer here is in charge of connection handling while the WorkerService implements the RPC and interfacing with the job package. I think separating responsibility like this will be easier for future implementation and expansion

Reply:
After reconsidering, I agree that it is better to combine the 2. They are now combined into 1 in the latest commit.